volatility/framework/symbols/windows/extensions/__init__.py

```diff
@@ -666,3 +700,83 @@ def to_list(self,
+                if sid == None:
+                    raise StopIteration
+                # this mimics the windows API IsValidSid
+                if sid.Revision & 0xF != 1 or sid.SubAuthorityCount > 15:
+                    raise StopIteration
+                id_auth = ""
+                for i in sid.IdentifierAuthority.Value:
+                    id_auth = i
+                SubAuthority = ntkrnlmp.object(object_type="array",
+                                               offset=sid.SubAuthority.vol.offset - kvo,
+                                               subtype = ntkrnlmp.get_type("unsigned long"),
+                                               count= int(sid.SubAuthorityCount))
+                yield "S-" + "-".join(str(i) for i in (sid.Revision, id_auth) +
+                                      tuple(SubAuthority))
+            
+
+    def privileges(self):
+        "Return a list of privileges for the current token object."
+        for priv_index in range(64):
+            yield (priv_index,
+                   self.Privileges.Present & (2**priv_index) !=0,
+                   self.Privileges.Enabled & (2**priv_index) !=0,
+                   self.Privileges.EnabledByDefault & (2**priv_index) !=0)
+
+
+class KTHREAD(objects.StructType):
+    """A class for thread control block objects."""
+        
+    def get_state(self) -> str:
+        dictState = {0:'Initialized', 1: 'Ready', 2: 'Running', 3: 'Standby', 4: 'Terminated',
+                    5: 'Waiting', 6: 'Transition', 7: 'DeferredReady', 8: 'GateWait'}
+        state = int(self.State)
+        stringState = ''
+        for flag in dictState:
+            if state == flag:
+                stringState += '{} '.format(dictState[flag])
+
+        return stringState[:-1] if len(stringState) else stringState
+
+    def get_wait_reason(self) -> str:
+        dictWaitReason = {0: 'Executive', 1: 'FreePage', 2: 'PageIn', 3: 'PoolAllocation',
+                    4: 'DelayExecution', 5: 'Suspended', 6: 'UserRequest', 7: 'WrExecutive',
+                    8: 'WrFreePage', 9: 'WrPageIn', 10: 'WrPoolAllocation', 11: 'WrDelayExecution',
+                    12: 'WrSuspended', 13: 'WrUserRequest', 14: 'WrEventPair', 15: 'WrQueue',
+                    16: 'WrLpcReceive', 17: 'WrLpcReply', 18: 'WrVirtualMemory', 19: 'WrPageOut',
+                    20: 'WrRendezvous', 21: 'Spare2', 22: 'Spare3', 23: 'Spare4', 24: 'Spare5',
+                    25: 'Spare6', 26: 'WrKernel', 27: 'WrResource', 28: 'WrPushLock', 29: 'WrMutex',
+                    30: 'WrQuantumEnd', 31: 'WrDispatchInt', 32: 'WrPreempted',
+                    33: 'WrYieldExecution', 34: 'WrFastMutex', 35: 'WrGuardedMutex',
+                    36: 'WrRundown', 37: 'MaximumWaitReason'}
+        waitReason = int(self.WaitReason)
+        stringWaitReason = ''
+        for flag in dictWaitReason:
+            if waitReason == flag:
+                stringWaitReason += '{} '.format(dictWaitReason[flag])
+
```

Comment:
@memoryforensics1 I am not sure I understand this loop over the hash table.

Since its a hash table, each integer key can only be present once, and it seems like your keys in 'dictWaitReason' are all you unique. Because of that, I am not sure why you loop over all of the element and attempt to allow for building of multiple values when only one could match.

I looked up the related Volatility 2 code and it seems to only just pull the one value:

https://gitlab.research.volexity.com/Volcano/volatility_volc/-/blob/master/volatility/plugins/malware/threads.py#L509

The code in your patch for get_state() does a similar pattern that I do not understand.

Reply:
My bad
I will close this and open another pull request